src/librustc_mir/shim.rs

```diff
@@ -842,29 +841,26 @@ fn build_call_shim<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     mir
 }
 
-pub fn build_adt_ctor<'a, 'gcx, 'tcx>(infcx: &infer::InferCtxt<'a, 'gcx, 'tcx>,
-                                      ctor_id: hir::HirId,
-                                      fields: &[hir::StructField],
-                                      span: Span)
-                                      -> Mir<'tcx>
-{
-    let tcx = infcx.tcx;
-    let gcx = tcx.global_tcx();
-    let def_id = tcx.hir().local_def_id_from_hir_id(ctor_id);
-    let param_env = gcx.param_env(def_id);
+pub fn build_adt_ctor<'a, 'gcx>(tcx: TyCtxt<'a, 'gcx, 'gcx>, ctor_id: DefId) -> &'gcx Mir<'gcx> {
```

Comment:
cc @rust-lang/lang 	Not suggesting that we add this or anything, but wanted to let you know that we can now generate a function

```rust
fn foo(a: A, b: B, c: C) -> Foo {
    Foo { a, b, c }
}
```

from a struct definition

```rust
struct Foo {
    a: A,
    b: B,
    c: C,
}
```

In case this ever comes up as something that may be useful

Reply:
@oli-obk FWIW we already have an optimization that expands `Rvalue::Aggregate` into this instruction sequence so this code is already duplicated, I think?

Reply:
iirc [that optimization](https://github.com/rust-lang/rust/blob/master/src/librustc_mir/transform/deaggregator.rs) is very expensive to run. I know we want to do little work at MIR building time, but in this case I believe it's equivalent in complexity. We can scrap that optimization if we just do it all at MIR building time.

Reply:
I meant that we could put the code generating this in that optimization until we get rid of it (rather than relying on running the optimization).